rotkehlchen/exchanges/kraken.py

```diff
@@ -1050,3 +1145,137 @@ def history_event_from_kraken(
             returned_events.append(event)
 
         return returned_events, skipped, found_unknown_event
+
+    #################
+    # Futures methods
+    #################
+
+    def _query_futures_api_method(self, method: str, req: dict | None = None) -> dict | str:
+        """API queries that require a valid key/secret pair.
+
+        Arguments:
+        method -- API method name (string, no default)
+        req    -- additional API request parameters (default: {})
+
+        """
+        if req is None:
+            req = {}
+
+        urlpath: str = '/derivatives/api/' + KRAKEN_FUTURES_API_VERSION + '/' + method if method is not None else ''  # noqa: E501
+        urlpath_without_prefix = urlpath.removeprefix('/derivatives')
+        req['nonce'] = str(int(1000 * time.time()))
+        post_data = ''
+
+        # any unicode strings must be turned to bytes
+        hashable = (post_data + req['nonce'] + urlpath_without_prefix).encode()
+        message = hashlib.sha256(hashable).digest()
+        signature = self.generate_hmac_b64_signature(
+            secret=self.futures_api_secret,
+            message=message,
+            digest_algorithm=hashlib.sha512,
+        )
+        self.session.headers.update({
+            'APIKey': typing.cast('ApiKey', self.futures_api_key),
+            'Nonce': req['nonce'],
+            'Authent': signature,
+        })
```

Comment:
actually this is wrong and something that is bad in our current implementation. So the scenario is two greenlets that are running at the same time:

1. greenlet 1 sets headers
2. greenlet 1 context switches before doing the request
3. greenlet 2 sets headers
4. greenlet 2 context switches before doing the requests
5. greenlet 1 sends request and signature is wrong

To fix this instead of updating the session headers use the `header` argument of `.get`. It should work